tools/mpremote/mpremote/repl.py

```diff
@@ -7,51 +7,53 @@ def do_repl_main_loop(
+            c = console_in.readchar()
+            if c:
+                if c in (b"\x1d", b"\x18"):  # ctrl-] or ctrl-x, quit
+                    break
+                elif c == b"\x04":  # ctrl-D
+                    # special handling needed for ctrl-D if filesystem is mounted
+                    state.transport.write_ctrl_d(console_out_write)
+                elif c == b"\x0a" and code_to_inject is not None:  # ctrl-j, inject code
+                    state.transport.serial.write(code_to_inject)
+                elif c == b"\x0b" and file_to_inject is not None:  # ctrl-k, inject script
+                    console_out_write(bytes("Injecting %s\r\n" % file_to_inject, "utf8"))
+                    state.transport.enter_raw_repl(soft_reset=False)
+                    with open(file_to_inject, "rb") as f:
+                        pyfile = f.read()
+                    try:
+                        state.transport.exec_raw_no_follow(pyfile)
+                    except TransportError as er:
+                        console_out_write(b"Error:\r\n")
+                        console_out_write(er)
+                    state.transport.exit_raw_repl()
+                else:
+                    state.transport.serial.write(c)
+
             n = state.transport.serial.inWaiting()
-        except OSError as er:
-            if er.args[0] == 5:  # IO error, device disappeared
-                print("device disconnected")
-                break
+            if n > 0:
+                dev_data_in = state.transport.serial.read(n)
+                if dev_data_in is not None:
+                    if escape_non_printable:
+                        # Pass data through to the console, with escaping of non-printables.
+                        console_data_out = bytearray()
+                        for c in dev_data_in:
+                            if c in (8, 9, 10, 13, 27) or 32 <= c <= 126:
+                                console_data_out.append(c)
+                            else:
+                                console_data_out.extend(b"[%02x]" % c)
+                    else:
+                        console_data_out = dev_data_in
+                    console_out_write(console_data_out)
 
-        if n > 0:
-            dev_data_in = state.transport.serial.read(n)
-            if dev_data_in is not None:
-                if escape_non_printable:
-                    # Pass data through to the console, with escaping of non-printables.
-                    console_data_out = bytearray()
-                    for c in dev_data_in:
-                        if c in (8, 9, 10, 13, 27) or 32 <= c <= 126:
-                            console_data_out.append(c)
-                        else:
-                            console_data_out.extend(b"[%02x]" % c)
-                else:
-                    console_data_out = dev_data_in
-                console_out_write(console_data_out)
+        except OSError as er:
```

Comment:
Re-looking at this now, I think it makes more sense to try/except the whole loop rather than a couple of individual checks. This way exceptions in the various transport read/write functions are also captured.

The only risk is of a OSError exception from console writes also showing as a device disconnection, I'm not sure if this is a particularly likely scenareo though?